### PR TITLE
SumState-Alerting

### PR DESCRIPTION
**Pull request by da-Kai**
## Description:

Notify specific users of the persistent presence of a negative sumState.

This pull request is linked to the following:
https://github.com/OpenEMS/openems/pull/2260

### CODE REVIEW

File: openems/controllers/alerting.py [hunk 1]
```diff
@@ -13,41 +18,75 @@ def __init__(self, sentAt: datetime, edgeId: str, userIds: list[int]) -> None:
...
+    def sum_state_alerting(self, sentAt: str, params: list[dict]):
+        msgs = self.__get_sum_state_params(sentAt, params)
+        update_func = lambda role, at: { role.write({"sum_state_last_notification": at})}
+        
+        if len(msgs) == 0:
+            self.__logger.error("Scheduled SumState-Alerting-Mail without any recipients!!!")
+        
+        for msg in msgs:
+            template = request.env.ref('openems.alerting_sum_state')
+            #template = self.__get_template(msg.edgeId)
```
Comment (michaelgrill) [inline openems/controllers/alerting.py]: remove unnecessary comments

File: openems/controllers/alerting.py [hunk 2]
```diff
@@ -13,41 +18,75 @@ def __init__(self, sentAt: datetime, edgeId: str, userIds: list[int]) -> None:
...
+        roles = http.request.env['openems.alerting'].search(
+            [('user_id','in',msg.userIds),('device_id','=',msg.edgeId)]
+        )
+        
         for role in roles:
             try:
                 template.send_mail(res_id=role.id, force_send=True)
-                role.write({"last_notification": msg.sentAt})
+                update_func(role, msg.sentAt)
+                #role.write({"offline_last_notification": msg.sentAt})
```
Comment (michaelgrill) [inline openems/controllers/alerting.py]: remove unnecessary comments

File: openems/mail/openems/alerting_sum_state.xml [hunk 3]
```diff
@@ -119,34 +117,35 @@
...
+                                <div style="font-family:helvetica;font-size:16px;line-height:1;text-align:left;color:#000000;">Your OpenEMS Edge with number <strong t-out="object.device_id.name"></strong> is in a continuous <t t-if="object.device_id.openems_sum_state_level">
+                                    <t t-set="sumstate" t-value="object.device_id.openems_sum_state_level.upper()">
+                                    </t>
+                                    <strong t-att-style="'color: red' if sumstate == 'FAULT' else 'color: yellow' if sumstate == 'WARNING' else ''" t-out="sumstate">
+                                    </strong>
                                   </t>
                                   <t t-else>
-                                    <t t-out="format_datetime(object.device_id.lastmessage, tz='UTC', dt_format='dd.MM.YYYY HH:mm:ss')"></t> (UTC)
-                                  </t>
+                                    <stong>UNKNOWN </stong>
```
Comment (michaelgrill) [inline openems/mail/openems/alerting_sum_state.xml]: i think you meant `<strong>`

File: openems/mail/openems/alerting_sum_state.xml [hunk 4]
```diff
@@ -212,12 +218,33 @@
                   <tbody>
                     <tr>
                       <td align="left" style="font-size:0px;padding:10px 25px;word-break:break-word;">
-                        <div style="font-family:helvetica;font-size:16px;line-height:1;text-align:left;color:#000000;">Wenn Sie keine E-Mail Benachrichtigung mehr wünschen, können Sie die Funktion <a class="link" href=" http://localhost:4200/device/${object.device_id.name}/settings/alerting" style="text-decoration: none; color: #36aed1;">hier</a> deaktivieren.</div>
+                        <div style="font-family:helvetica;font-size:16px;line-height:1;text-align:left;color:#000000;">If you do no longer wish to receive email notifications, you can disable the feature <a class="link" href="http://localhost:4200/device/${object.device_id.name}/settings/alerting" style="text-decoration: none; color: #e8ae6a;"> hier </a>.</div>
```
Comment (michaelgrill) [inline openems/mail/openems/alerting_sum_state.xml]: you forgot to translate `hier` to `here` at the and with the link

File: openems/controllers/alerting.py [hunk 5]
```diff
@@ -13,41 +18,75 @@ def __init__(self, sentAt: datetime, edgeId: str, userIds: list[int]) -> None:
...
+    @http.route("/openems_backend/mail/alerting_sum_state", type="json", auth="user")
+    def sum_state_alerting(self, sentAt: str, params: list[dict]):
+        msgs = self.__get_sum_state_params(sentAt, params)
+        update_func = lambda role, at: { role.write({"sum_state_last_notification": at})}
+        
+        if len(msgs) == 0:
+            self.__logger.error("Scheduled SumState-Alerting-Mail without any recipients!!!")
+        
+        for msg in msgs:
+            template = request.env.ref('openems.alerting_sum_state')
```
Comment (michaelgrill) [inline openems/controllers/alerting.py]: wouldnt it make more sense to get this outsite the loop?

File: openems/controllers/alerting.py [hunk 6]
```diff
@@ -13,41 +18,75 @@ def __init__(self, sentAt: datetime, edgeId: str, userIds: list[int]) -> None:
...
-            msgs.append(Message(sent, param["edgeId"], param["recipients"]));
+            edgeId = param["edgeId"]
+            recipients = param["recipients"]
+            state = param["state"]
+            msgs.append(SumStateMessage(sent, edgeId, recipients, state));
         return msgs
 
     def __get_template(self, device_id):
-        template = request.env.ref("openems.alerting_email_generic")
+        template = request.env.ref("openems.alerting_offline")
```
Comment (michaelgrill) [inline openems/controllers/alerting.py]: do you even need the `device_id` here? If not you could also move the method call outside the loop; also maybe decide on one way to get the template via method or directly with env

File: openems/models/device.py [hunk 7]
```diff
@@ -174,3 +175,50 @@ def _compute_text_teaser(self):
...
+    def _compute_device_name(self):
+        for rec in self:
+            rec.device_name = rec.device_id.name;
+
+    @api.depends("user_id")
+    def _compute_user_login(self):
+        for rec in self:
+            rec.user_login = rec.user_id.login;
+
+    @api.depends("user_id")
```
Comment (michaelgrill) [inline openems/models/device.py]: does it also need to depend on `device_id`?

**Review by michaelgrill — CHANGES_REQUESTED** [review]

**Review by michaelgrill — APPROVED** [review]
🙂👍

**Comment by michaelgrill** [comment]
@sfeilmeier FYI finished with my review

**Review by sfeilmeier — APPROVED** [review]
Ok from my side!

Please do an internal PR as well, because Odoo-OpenEMS-Addon and our internal module are not fully in sync yet, unfortunately.

**Pull request merged**